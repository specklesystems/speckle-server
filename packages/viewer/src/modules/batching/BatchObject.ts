/* eslint-disable camelcase */
import { Box3, Euler, Matrix4, Quaternion, Vector3 } from 'three'
import { NodeRenderView } from '../tree/NodeRenderView'
import { Geometry } from '../converter/Geometry'
import {
  AccelerationStructure,
  DefaultBVHOptions
} from '../objects/AccelerationStructure'
import { MeshBVH } from 'three-mesh-bvh'

export type VectorLike =
  | { x: number; y: number; z?: number; w?: number }
  | undefined
  | null
export type Vector3Like = VectorLike & { z: number }
export type Vector4Like = Vector3Like & { w: number }

export class BatchObject {
  protected _renderView: NodeRenderView
  protected _accelerationStructure: AccelerationStructure
  protected _batchIndex: number
  protected _localOrigin: Vector3
  public transform: Matrix4
  public transformInv: Matrix4

  public tasVertIndexStart!: number
  public tasVertIndexEnd!: number

  public quaternion: Quaternion = new Quaternion()
  public eulerValue: Euler = new Euler()
  public pivot_High: Vector3 = new Vector3()
  public pivot_Low: Vector3 = new Vector3()
  public translation: Vector3 = new Vector3()
  public scaleValue: Vector3 = new Vector3(1, 1, 1)

  protected static matBuff0: Matrix4 = new Matrix4()
  protected static matBuff1: Matrix4 = new Matrix4()
  protected static matBuff2: Matrix4 = new Matrix4()
  protected static eulerBuff: Euler = new Euler()
  protected static translationBuff: Vector3 = new Vector3()
  protected static scaleBuff: Vector3 = new Vector3()
  protected static pivotBuff: Vector3 = new Vector3()

  public transformDirty = true

  public get renderView(): NodeRenderView {
    return this._renderView
  }

  public get accelerationStructure(): AccelerationStructure {
    return this._accelerationStructure
  }

  public get batchIndex(): number {
    return this._batchIndex
  }

  public get aabb(): Box3 {
    if (this.renderView.aabb) {
      const box = new Box3().copy(this.renderView.aabb)
      box.applyMatrix4(this.transform)
      return box
    }
    return new Box3()
  }

  public get localOrigin(): Vector3 {
    return this._localOrigin
  }

  public set position(value: Vector3) {
    this.transformTRS(
      new Vector3().subVectors(value, this._localOrigin),
      this.eulerValue,
      this.scaleValue,
      new Vector3().addVectors(this.pivot_Low, this.pivot_High)
    )
  }

  public set euler(euler: Euler) {
    this.transformTRS(
      this.translation,
      euler,
      this.scaleValue,
      new Vector3().addVectors(this.pivot_Low, this.pivot_High)
    )
  }

  public set scale(scale: Vector3) {
    this.transformTRS(
      this.translation,
      this.eulerValue,
      scale,
      new Vector3().addVectors(this.pivot_Low, this.pivot_High)
    )
  }

  public constructor(renderView: NodeRenderView, batchIndex: number) {
    this._renderView = renderView
    this._batchIndex = batchIndex
    this.transform = new Matrix4().identity()
    this.transformInv = new Matrix4().identity()

    this._localOrigin = this._renderView.aabb.getCenter(new Vector3())
    Geometry.DoubleToHighLowVector(
      new Vector3(this._localOrigin.x, this._localOrigin.y, this._localOrigin.z),
      this.pivot_Low,
      this.pivot_High
    )
  }

  public buildAccelerationStructure(bvh?: MeshBVH) {
    const transform = new Matrix4().makeTranslation(
      this._localOrigin.x,
      this._localOrigin.y,
      this._localOrigin.z
    )
    transform.invert()

    if (!bvh) {
      const indices: number[] | undefined =
        this._renderView.renderData.geometry.attributes?.INDEX
      const position: number[] | undefined =
        this._renderView.renderData.geometry.attributes?.POSITION
      bvh = AccelerationStructure.buildBVH(
        indices,
        position,
        DefaultBVHOptions,
        transform
      )
    }

    this._accelerationStructure = new AccelerationStructure(bvh)
    this._accelerationStructure.inputTransform = this.transformInv
    this._accelerationStructure.outputTransform = this.transform
    this._accelerationStructure.inputOriginTransform = new Matrix4().copy(transform)
    this._accelerationStructure.outputOriginTransfom = new Matrix4()
      .copy(transform)
      .invert()
  }

  public transformTRS(
    translation: Vector3Like,
    euler?: Vector3Like,
    scale?: Vector3Like,
    pivot?: Vector3Like
  ) {
    let T: Matrix4 = BatchObject.matBuff0.identity()
    let R: Matrix4 = BatchObject.matBuff1.identity()
    let S: Matrix4 = BatchObject.matBuff2.identity()
    BatchObject.eulerBuff.set(0, 0, 0, 'XYZ')
    BatchObject.translationBuff.set(0, 0, 0)
    BatchObject.scaleBuff.set(1, 1, 1)
    BatchObject.pivotBuff.copy(this._localOrigin)

    if (translation) {
      T = BatchObject.matBuff0.makeTranslation(
        translation.x,
        translation.y,
        translation.z
      )
      BatchObject.translationBuff.set(translation.x, translation.y, translation.z)
    }

    if (euler) {
      BatchObject.eulerBuff.set(euler.x, euler.y, euler.z, 'XYZ')
      R = BatchObject.matBuff1.makeRotationFromEuler(BatchObject.eulerBuff)
      this.quaternion.setFromEuler(BatchObject.eulerBuff)
      this.eulerValue.copy(BatchObject.eulerBuff)
    }

    if (scale) {
      S = BatchObject.matBuff2.makeScale(scale.x, scale.y, scale.z)
      BatchObject.scaleBuff.set(scale.x, scale.y, scale.z)
    }

    if (pivot) {
      BatchObject.pivotBuff.set(pivot.x, pivot.y, pivot.z)
    }

    this.transform.identity()
    this.transform.multiply(R)
    this.transform.multiply(S)
    this.transform.premultiply(T)

    this.transformInv.copy(this.transform)
    this.transformInv.invert()

    this.translation.copy(BatchObject.translationBuff)
    this.quaternion.setFromEuler(BatchObject.eulerBuff)
    this.scaleValue.copy(BatchObject.scaleBuff)

    Geometry.DoubleToHighLowVector(
      BatchObject.pivotBuff,
      this.pivot_Low,
      this.pivot_High
    )
    this.transformDirty = true
  }
}
