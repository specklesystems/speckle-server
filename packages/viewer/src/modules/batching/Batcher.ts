import { generateUUID } from 'three/src/math/MathUtils'
import MeshBatch from './MeshBatch'
import { SpeckleType } from '../converter/GeometryConverter'
import { WorldTree } from '../tree/WorldTree'
import LineBatch from './LineBatch'
import Materials from '../materials/Materials'
import { NodeRenderView } from '../tree/NodeRenderView'
import {
  AllBatchUpdateRange,
  Batch,
  BatchUpdateRange,
  GeometryType,
  HideAllBatchUpdateRange
} from './Batch'
import PointBatch from './PointBatch'
// import { FilterMaterialType } from '../FilteringManager'
import { Material, Mesh, WebGLRenderer } from 'three'
import { FilterMaterial, FilterMaterialType } from '../filtering/FilteringManager'
import Logger from 'js-logger'

export default class Batcher {
  public materials: Materials
  public batches: { [id: string]: Batch } = {}

  public constructor() {
    this.materials = new Materials()
    this.materials.createDefaultMaterials()
  }

  public makeBatches(
    subtreeId: string,
    batchType: GeometryType,
    ...speckleType: SpeckleType[]
  ) {
    const rendeViews = WorldTree.getRenderTree(subtreeId)
      .getAtomicRenderViews(...speckleType)
      .sort((a, b) => {
        if (a.renderMaterialHash === 0) return -1
        if (b.renderMaterialHash === 0) return 1
        return a.renderMaterialHash - b.renderMaterialHash
      })
    const materialHashes = [
      ...Array.from(new Set(rendeViews.map((value) => value.renderMaterialHash)))
    ]

    Logger.warn(materialHashes)
    // console.warn(rendeViews)

    for (let i = 0; i < materialHashes.length; i++) {
      const batch = rendeViews.filter(
        (value) => value.renderMaterialHash === materialHashes[i]
      )

      let matRef = null

      if (batchType === GeometryType.MESH) {
        matRef = batch[0].renderData.renderMaterial
      } else if (batchType === GeometryType.LINE) {
        matRef = batch[0].renderData.displayStyle
      } else if (batchType === GeometryType.POINT) {
        matRef = batch[0].renderData.renderMaterial
      } else if (batchType === GeometryType.POINT_CLOUD) {
        matRef = batch[0].renderData.renderMaterial
      }

      const material = this.materials.updateMaterialMap(
        materialHashes[i],
        matRef,
        batchType
      )

      const batchID = generateUUID()
      switch (batchType) {
        case GeometryType.MESH:
          this.batches[batchID] = new MeshBatch(batchID, subtreeId, batch)
          break
        case GeometryType.LINE:
          this.batches[batchID] = new LineBatch(batchID, subtreeId, batch)
          break
        case GeometryType.POINT:
          this.batches[batchID] = new PointBatch(batchID, subtreeId, batch)
          break
        case GeometryType.POINT_CLOUD:
          this.batches[batchID] = new PointBatch(batchID, subtreeId, batch)
          break
      }

      this.batches[batchID].setBatchMaterial(material)
      this.batches[batchID].buildBatch()
      // console.warn(batch)
    }
  }

  public update(deltaTime: number) {
    for (const batchId in this.batches) {
      this.batches[batchId].onUpdate(deltaTime)
    }
  }

  public render(renderer: WebGLRenderer) {
    for (const batchId in this.batches) {
      this.batches[batchId].onRender(renderer)
    }
  }

  public saveVisiblity(): Record<string, BatchUpdateRange> {
    const visibilityRanges = {}
    for (const k in this.batches) {
      const batch: Batch = this.batches[k]
      // if (batch.geometryType !== GeometryType.MESH) continue
      visibilityRanges[k] = batch.getVisibleRange()
    }
    return visibilityRanges
  }

  public applyVisibility(ranges: Record<string, BatchUpdateRange>) {
    for (const k in this.batches) {
      const batch: Batch = this.batches[k]
      // if (batch.geometryType !== GeometryType.MESH) continue
      const range = ranges[k]
      if (!range) {
        batch.setVisibleRange(HideAllBatchUpdateRange)
      } else {
        batch.setVisibleRange(range)
      }
    }
  }

  public getTransparent(): Record<string, BatchUpdateRange> {
    const visibilityRanges = {}
    for (const k in this.batches) {
      const batch: Batch = this.batches[k]
      if (batch.geometryType !== GeometryType.MESH) {
        visibilityRanges[k] = HideAllBatchUpdateRange
        continue
      }
      const batchMesh: Mesh = batch.renderObject as Mesh
      if (batchMesh.geometry.groups.length === 0) {
        if ((batchMesh.material as Material).transparent === true)
          visibilityRanges[k] = AllBatchUpdateRange
      } else {
        const transparentGroup = batchMesh.geometry.groups.find((value) => {
          return batchMesh.material[value.materialIndex].visible === true
        })
        const hiddenGroup = batchMesh.geometry.groups.find((value) => {
          return batchMesh.material[value.materialIndex].visible === false
        })
        if (transparentGroup) {
          visibilityRanges[k] = {
            offset: transparentGroup.start,
            count:
              hiddenGroup !== undefined
                ? hiddenGroup.start
                : batch.getCount() - transparentGroup.start
          }
        }
      }
    }
    return visibilityRanges
  }

  public getOpaque() {
    const visibilityRanges = {}
    for (const k in this.batches) {
      const batch: Batch = this.batches[k]
      if (batch.geometryType !== GeometryType.MESH) {
        visibilityRanges[k] = HideAllBatchUpdateRange
        continue
      }
      const batchMesh: Mesh = batch.renderObject as Mesh
      if (batchMesh.geometry.groups.length === 0) {
        if ((batchMesh.material as Material).transparent === false)
          visibilityRanges[k] = AllBatchUpdateRange
      } else {
        const transparentOrHiddenGroup = batchMesh.geometry.groups.find((value) => {
          return (
            batchMesh.material[value.materialIndex].transparent === true ||
            batchMesh.material[value.materialIndex].visible === false
          )
        })
        visibilityRanges[k] = {
          offset: 0,
          count:
            transparentOrHiddenGroup !== undefined
              ? transparentOrHiddenGroup.start
              : batch.getCount()
        }
      }
    }
    return visibilityRanges
  }

  public enableTransparent(value: boolean) {
    for (const k in this.batches) {
      const batch: Batch = this.batches[k]
      if (batch.geometryType !== GeometryType.MESH) continue
      const batchMesh: Mesh = batch.renderObject as Mesh
      if (batchMesh.geometry.groups.length === 0) {
        batchMesh.visible = (batchMesh.material as Material).transparent
          ? value
          : batchMesh.visible
      } else {
        const transparentGroup = batchMesh.geometry.groups.find((value) => {
          return batchMesh.material[value.materialIndex].transparent === true
        })
        batch.setVisibleRange({
          offset: 0,
          count: transparentGroup.start
        })
      }
    }
  }

  public purgeBatches(subtreeId: string) {
    for (const k in this.batches) {
      if (this.batches[k].subtreeId === subtreeId) {
        this.batches[k].purge()
        delete this.batches[k]
      }
    }
  }

  public getBatches(subtreeId?: string, geometryType?: GeometryType) {
    return Object.values(this.batches).filter((value: Batch) => {
      const subtree = subtreeId !== undefined ? value.subtreeId === subtreeId : true
      const type =
        geometryType !== undefined ? value.geometryType === geometryType : true
      return subtree && type
    })
  }

  public getRenderView(batchId: string, index: number) {
    return this.batches[batchId].getRenderView(index)
  }

  public resetBatchesDrawRanges() {
    for (const k in this.batches) {
      this.batches[k].resetDrawRanges()
    }
  }

  public setObjectsFilterMaterial(
    rvs: NodeRenderView[],
    filterMaterial: FilterMaterial,
    uniqueRvsOnly = true
  ): string[] {
    let renderViews = rvs
    if (uniqueRvsOnly) renderViews = [...Array.from(new Set(rvs.map((value) => value)))]
    const batchIds = [...Array.from(new Set(renderViews.map((value) => value.batchId)))]
    for (let i = 0; i < batchIds.length; i++) {
      if (!batchIds[i]) {
        continue
      }
      const batch = this.batches[batchIds[i]]
      const views = renderViews
        .filter((value) => value.batchId === batchIds[i])
        .map((rv: NodeRenderView) => {
          return {
            offset: rv.batchStart,
            count: rv.batchCount,
            material: this.materials.getFilterMaterial(
              rv,
              filterMaterial.filterType
            ) as Material,
            materialOptions: this.materials.getFilterMaterialOptions(filterMaterial)
          } as BatchUpdateRange
        })
      batch.setDrawRanges(...views)
    }
    return batchIds
  }

  public autoFillDrawRanges(batchIds: string[]) {
    const uniqueBatches = [...Array.from(new Set(batchIds.map((value) => value)))]
    uniqueBatches.forEach((value) => {
      if (!value) return
      this.batches[value].autoFillDrawRanges()
    })
  }

  /** Conveniece method. This should also work as a filtering action
   *  Though, because the batches are not smart enough yet to group
   *  their draw ranges, it would be currently be inneficient to isolate
   *  via filtering. This will change in the future
   */
  public isolateRenderView(id: string) {
    const rvs = WorldTree.getRenderTree().getRenderViewsForNodeId(id)
    const batchIds = [...Array.from(new Set(rvs.map((value) => value.batchId)))]
    for (const k in this.batches) {
      if (!batchIds.includes(k)) {
        this.batches[k].setDrawRanges({
          offset: 0,
          count: Infinity,
          material: this.materials.getFilterMaterial(
            this.batches[k].renderViews[0],
            FilterMaterialType.GHOST
          )
        })
        this.batches[k].setVisibleRange(HideAllBatchUpdateRange)
      } else {
        const drawRanges = []
        for (let i = 0; i < this.batches[k].renderViews.length; i++) {
          if (!rvs.includes(this.batches[k].renderViews[i])) {
            drawRanges.push({
              offset: this.batches[k].renderViews[i].batchStart,
              count: this.batches[k].renderViews[i].batchCount,
              material: this.materials.getFilterMaterial(
                this.batches[k].renderViews[i],
                FilterMaterialType.GHOST
              )
            })
          } else {
            drawRanges.push({
              offset: this.batches[k].renderViews[i].batchStart,
              count: this.batches[k].renderViews[i].batchCount,
              material: this.materials.getFilterMaterial(
                this.batches[k].renderViews[i],
                FilterMaterialType.SELECT
              )
            })
          }
        }
        if (drawRanges.length > 0) {
          this.batches[k].setDrawRanges(...drawRanges)
          this.batches[k].autoFillDrawRanges()
        }
      }
    }
  }

  public async isolateRenderViewBatch(id: string) {
    const rv = WorldTree.getRenderTree().getRenderViewForNodeId(id)
    for (const k in this.batches) {
      if (k !== rv.batchId) {
        this.batches[k].setDrawRanges({
          offset: 0,
          count: Infinity,
          material: this.materials.getFilterMaterial(
            this.batches[k].renderViews[0],
            FilterMaterialType.GHOST
          )
        })
      }
    }
  }
}
