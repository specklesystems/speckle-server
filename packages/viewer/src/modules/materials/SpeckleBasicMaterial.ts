/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable camelcase */
import { speckleBasicVert } from './shaders/speckle-basic-vert'
import { speckleBasicFrag } from './shaders/speckle-basic-frag'
import { UniformsUtils, ShaderLib, Vector3, MeshBasicMaterial } from 'three'
import { Matrix4 } from 'three'
import { Geometry } from '../converter/Geometry'

class SpeckleBasicMaterial extends MeshBasicMaterial {
  protected static readonly matBuff: Matrix4 = new Matrix4()
  protected static readonly vecBuff0: Vector3 = new Vector3()
  protected static readonly vecBuff1: Vector3 = new Vector3()
  protected static readonly vecBuff2: Vector3 = new Vector3()

  constructor(parameters, defines = []) {
    super(parameters)

    this.userData.uViewer_high = {
      value: new Vector3()
    }
    this.userData.uViewer_low = {
      value: new Vector3()
    }
    ;(this as any).vertProgram = speckleBasicVert
    ;(this as any).fragProgram = speckleBasicFrag
    ;(this as any).uniforms = UniformsUtils.merge([
      ShaderLib.standard.uniforms,
      {
        uViewer_high: {
          value: this.userData.uViewer_high.value
        },
        uViewer_low: {
          value: this.userData.uViewer_low.value
        }
      }
    ])

    this.onBeforeCompile = function (shader) {
      shader.uniforms.uViewer_high = this.userData.uViewer_high
      shader.uniforms.uViewer_low = this.userData.uViewer_low
      shader.vertexShader = this.vertProgram
      shader.fragmentShader = this.fragProgram
    }

    if (defines) {
      this.defines = {}
    }
    for (let k = 0; k < defines.length; k++) {
      this.defines[defines[k]] = ' '
    }
  }

  copy(source) {
    super.copy(source)
    this.userData = {}
    this.userData.uViewer_high = {
      value: new Vector3()
    }
    this.userData.uViewer_low = {
      value: new Vector3()
    }

    this.defines['USE_RTE'] = ' '

    return this
  }

  onBeforeRender(_this, scene, camera, geometry, object, group) {
    SpeckleBasicMaterial.matBuff.copy(camera.matrixWorldInverse)
    SpeckleBasicMaterial.matBuff.elements[12] = 0
    SpeckleBasicMaterial.matBuff.elements[13] = 0
    SpeckleBasicMaterial.matBuff.elements[14] = 0
    SpeckleBasicMaterial.matBuff.multiply(object.matrixWorld)
    object.modelViewMatrix.copy(SpeckleBasicMaterial.matBuff)

    SpeckleBasicMaterial.vecBuff0.set(
      camera.matrixWorld.elements[12],
      camera.matrixWorld.elements[13],
      camera.matrixWorld.elements[14]
    )

    Geometry.DoubleToHighLowVector(
      SpeckleBasicMaterial.vecBuff0,
      SpeckleBasicMaterial.vecBuff1,
      SpeckleBasicMaterial.vecBuff2
    )

    this.userData.uViewer_low.value.copy(SpeckleBasicMaterial.vecBuff1)
    this.userData.uViewer_high.value.copy(SpeckleBasicMaterial.vecBuff2)

    this.needsUpdate = true
  }
}

export default SpeckleBasicMaterial
