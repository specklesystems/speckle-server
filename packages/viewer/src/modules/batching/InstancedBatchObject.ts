/* eslint-disable camelcase */
import { BatchObject, type Vector3Like } from './BatchObject'

import { Matrix4 } from 'three'
import { NodeRenderView } from '../tree/NodeRenderView'

export class InstancedBatchObject extends BatchObject {
  protected instanceTransform: Matrix4 = new Matrix4()

  public constructor(renderView: NodeRenderView, batchIndex: number) {
    super(renderView, batchIndex)
    if (renderView.renderData.geometry.transform)
      this.instanceTransform.copy(renderView.renderData.geometry.transform)
    this.transform.copy(this.instanceTransform)
    this.transformInv.copy(new Matrix4().copy(this.instanceTransform).invert())
    this.transformDirty = false
  }

  public transformTRS(
    translation: Vector3Like,
    euler: Vector3Like,
    scale: Vector3Like,
    pivot: Vector3Like
  ) {
    super.transformTRS(translation, euler, scale, pivot)
    this.transform.multiply(this.instanceTransform)
    this.transformInv.copy(this.transform)
    this.transformInv.invert()
  }
}
