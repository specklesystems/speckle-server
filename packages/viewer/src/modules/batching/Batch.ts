import { Material, Object3D, WebGLRenderer } from 'three'
import { MaterialOptions } from '../materials/Materials'
import { NodeRenderView } from '../tree/NodeRenderView'

export enum GeometryType {
  MESH,
  LINE,
  POINT,
  POINT_CLOUD
}

export interface Batch {
  id: string
  subtreeId: string
  renderViews: NodeRenderView[]
  batchMaterial: Material
  renderObject: Object3D
  geometryType: GeometryType

  getCount(): number
  setBatchMaterial(material: Material): void
  setVisibleRange(...range: BatchUpdateRange[])
  getVisibleRange(): BatchUpdateRange
  setDrawRanges(...ranges: BatchUpdateRange[])
  autoFillDrawRanges()
  resetDrawRanges()
  buildBatch()
  getRenderView(index: number): NodeRenderView
  onUpdate(deltaTime: number)
  onRender(renderer: WebGLRenderer)
  purge()
}

export interface BatchUpdateRange {
  offset: number
  count: number
  material?: Material
  materialOptions?: MaterialOptions
}

export const HideAllBatchUpdateRange = {
  offset: 0,
  count: 0
} as BatchUpdateRange

export const AllBatchUpdateRange = {
  offset: 0,
  count: Infinity
} as BatchUpdateRange
