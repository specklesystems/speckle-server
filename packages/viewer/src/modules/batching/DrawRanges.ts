import { Material } from 'three'
import { type BatchUpdateRange } from './Batch'
import { type DrawGroup } from './Batch'

export class DrawRanges {
  public integrateRanges(
    groups: Array<DrawGroup>,
    materials: Array<Material>,
    ranges: BatchUpdateRange[]
  ): Array<DrawGroup> {
    let _flatRanges: Array<number> = []
    groups.sort((a, b) => a.start - b.start)
    ranges.sort((a, b) => a.offset - b.offset)

    const edgesForward: { [key: number]: number } = {}
    const edgesBackwards: { [key: number]: number } = {}
    for (let k = 0, l = groups.length - 1; k < groups.length; k++, l--) {
      const groupForward = groups[k]
      const groupBackwards = groups[l]
      edgesForward[groupForward.start] = groupForward.materialIndex as number
      edgesBackwards[groupBackwards.start + groupBackwards.count] =
        groupBackwards.materialIndex as number
    }

    _flatRanges = groups.map((group: DrawGroup) => {
      return group.start + group.count
    })
    _flatRanges.unshift(0)

    for (let k = 0; k < ranges.length; k++) {
      const range = ranges[k]
      const r0 = range.offset
      const r0Index = _flatRanges.findIndex((value) => value > r0)
      const next = _flatRanges[r0Index]
      _flatRanges.splice(r0Index, 0, r0)

      const r1 = range.offset + range.count
      const r1Index = _flatRanges.findIndex((value) => value > r1)
      _flatRanges.splice(r1Index, 0, r1)

      _flatRanges = _flatRanges.filter((value) => {
        return !(value > r0 && value < r1)
      })
      _flatRanges = [...new Set(_flatRanges)]

      const materialIndex = materials.indexOf(range.material as Material)
      edgesForward[r0] = materialIndex
      edgesForward[r1] = r1 >= next ? edgesForward[next] : edgesBackwards[next]
    }

    const drawRanges = []
    let groupStart = -1
    let count = 0
    for (let k = 0; k < _flatRanges.length - 1; k++) {
      const start = _flatRanges[k]
      const end = _flatRanges[k + 1]
      count += end - start
      const materialIndex = edgesForward[start]
      const lastGroup = k === _flatRanges.length - 2
      if (edgesForward[_flatRanges[k + 1]] === materialIndex || lastGroup) {
        if (groupStart === -1) {
          groupStart = start
        }
        if (!lastGroup) continue
      }

      drawRanges.push({
        start: groupStart === -1 ? start : groupStart,
        count,
        materialIndex
      })
      groupStart = -1
      count = 0
    }

    return drawRanges
  }
}
