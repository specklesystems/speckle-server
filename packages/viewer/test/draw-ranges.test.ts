import { expect, describe, it } from 'vitest'
import { DrawRanges } from '../src/modules/batching/DrawRanges'
import SpeckleBasicMaterial from '../src/modules/materials/SpeckleBasicMaterial'
import { DrawGroup } from '../src/modules/batching/Batch'

const material0 = new SpeckleBasicMaterial({ color: 0xff0000 })
const material1 = new SpeckleBasicMaterial({ color: 0x00ff00 })
const material2 = new SpeckleBasicMaterial({ color: 0x0000ff })

describe('Draw Ranges', () => {
  it('Boundary Ranges', () => {
    const drawRange = new DrawRanges()
    let groups = [
      {
        start: 0,
        count: 2095,
        materialIndex: 0
      } as DrawGroup
    ]
    groups = drawRange.integrateRanges(
      groups,
      [material0, material1],
      [
        {
          offset: 0,
          count: 2095,
          material: material1
        }
      ]
    )
    expect(groups).toMatchSnapshot()

    groups = drawRange.integrateRanges(
      groups,
      [material0, material1],
      [
        {
          offset: 0,
          count: 2094,
          material: material0
        }
      ]
    )
    expect(groups).toMatchSnapshot()

    groups = [
      {
        start: 0,
        count: 2095,
        materialIndex: 0
      } as DrawGroup
    ]
    groups = drawRange.integrateRanges(
      groups,
      [material0, material1],
      [
        {
          offset: 2094,
          count: 1,
          material: material1
        }
      ]
    )
    expect(groups).toMatchSnapshot()
  })

  it('Mixed Ranges', () => {
    let groups = [
      {
        start: 0,
        count: 216,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 216,
        count: 1323,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 1539,
        count: 540,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 2079,
        count: 32268,
        materialIndex: 0
      } as DrawGroup
    ]

    const drawRange = new DrawRanges()

    groups = drawRange.integrateRanges(
      groups,
      [material0, material1],
      [
        { offset: 36, count: 36, material: material1 },
        {
          offset: 180,
          count: 1395,
          material: material1
        },
        {
          offset: 1581,
          count: 32766,
          material: material1
        }
      ]
    )
    expect(groups).toMatchSnapshot()
  })

  it('Mixed Ranges 2', () => {
    let groups = [
      {
        start: 0,
        count: 528,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 528,
        count: 3261,
        materialIndex: 1
      } as DrawGroup
    ]

    const drawRange = new DrawRanges()

    groups = drawRange.integrateRanges(
      groups,
      [material0, material1, material2],
      [{ offset: 6, count: 36, material: material2 }]
    )
    expect(groups).toMatchSnapshot()
  })

  it('Mixed Ranges 3', () => {
    let groups = [
      {
        start: 0,
        count: 528,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 528,
        count: 3261,
        materialIndex: 1
      } as DrawGroup
    ]

    const drawRange = new DrawRanges()

    groups = drawRange.integrateRanges(
      groups,
      [material0, material1, material2],
      [{ offset: 530, count: 36, material: material2 }]
    )
    expect(groups).toMatchSnapshot()
  })

  it('Multiple Materials', () => {
    const material3 = new SpeckleBasicMaterial({ color: 0x0000ff })
    const material4 = new SpeckleBasicMaterial({ color: 0x0000ff })

    let groups = [
      {
        start: 0,
        count: 216,
        materialIndex: 3
      } as DrawGroup,
      {
        start: 216,
        count: 1323,
        materialIndex: 1
      } as DrawGroup,
      {
        start: 1539,
        count: 540,
        materialIndex: 0
      } as DrawGroup,
      {
        start: 2079,
        count: 32268,
        materialIndex: 2
      } as DrawGroup
    ]

    const drawRange = new DrawRanges()

    groups = drawRange.integrateRanges(
      groups,
      [material0, material1, material2, material3, material4],
      [
        {
          offset: 36,
          count: 257,
          material: material4
        }
      ]
    )
    expect(groups).toMatchSnapshot()
  })
})
