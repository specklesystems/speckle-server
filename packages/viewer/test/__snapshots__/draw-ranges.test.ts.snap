// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Draw Ranges > Boundary Ranges 1`] = `
[
  {
    "count": 2095,
    "materialIndex": 1,
    "start": 0,
  },
]
`;

exports[`Draw Ranges > Boundary Ranges 2`] = `
[
  {
    "count": 2094,
    "materialIndex": 0,
    "start": 0,
  },
  {
    "count": 1,
    "materialIndex": 1,
    "start": 2094,
  },
]
`;

exports[`Draw Ranges > Boundary Ranges 3`] = `
[
  {
    "count": 2094,
    "materialIndex": 0,
    "start": 0,
  },
  {
    "count": 1,
    "materialIndex": 1,
    "start": 2094,
  },
]
`;

exports[`Draw Ranges > Mixed Ranges 1`] = `
[
  {
    "count": 36,
    "materialIndex": 0,
    "start": 0,
  },
  {
    "count": 36,
    "materialIndex": 1,
    "start": 36,
  },
  {
    "count": 108,
    "materialIndex": 0,
    "start": 72,
  },
  {
    "count": 1395,
    "materialIndex": 1,
    "start": 180,
  },
  {
    "count": 6,
    "materialIndex": 0,
    "start": 1575,
  },
  {
    "count": 32766,
    "materialIndex": 1,
    "start": 1581,
  },
]
`;

exports[`Draw Ranges > Mixed Ranges 2 1`] = `
[
  {
    "count": 6,
    "materialIndex": 0,
    "start": 0,
  },
  {
    "count": 36,
    "materialIndex": 2,
    "start": 6,
  },
  {
    "count": 486,
    "materialIndex": 0,
    "start": 42,
  },
  {
    "count": 3261,
    "materialIndex": 1,
    "start": 528,
  },
]
`;

exports[`Draw Ranges > Mixed Ranges 3 1`] = `
[
  {
    "count": 528,
    "materialIndex": 0,
    "start": 0,
  },
  {
    "count": 2,
    "materialIndex": 1,
    "start": 528,
  },
  {
    "count": 36,
    "materialIndex": 2,
    "start": 530,
  },
  {
    "count": 3223,
    "materialIndex": 1,
    "start": 566,
  },
]
`;

exports[`Draw Ranges > Multiple Materials 1`] = `
[
  {
    "count": 36,
    "materialIndex": 3,
    "start": 0,
  },
  {
    "count": 257,
    "materialIndex": 4,
    "start": 36,
  },
  {
    "count": 1246,
    "materialIndex": 1,
    "start": 293,
  },
  {
    "count": 540,
    "materialIndex": 0,
    "start": 1539,
  },
  {
    "count": 32268,
    "materialIndex": 2,
    "start": 2079,
  },
]
`;
